rep-2012.rst

```diff
@@ -0,0 +1,466 @@
+
+There are additional features that could leverage this proposal, such as:
+
+- Playback of recorded services (for example, from a rosbag [1]_)
+- Introspection of ROS actions, which are built on services
+- Validation of a live ROS system by referencing a recording from a previous session
+
+Though this proposal focuses on the core feature of introspecting requests and responses, we want to keep the design flexible so additional features like those listed above can be implemented in the future.
+
+
+Specification
+=============
+
+Publishing Service Events
+-------------------------
+
+Whenever a request or response is sent or received, a *service event* message will be published to a topic.
+Services are responsible for publishing a message when they receive a request and when they send a response.
+Likewise, clients are responsible for publishing a message when they send a request and when they receive a response.
+Therefore, we have a total of four possible events:
+
+:Request Sent:
+  Emitted from a client after sending a request to a service.
+:Request Received:
+  Emitted from a service after receiving a request from a client.
+:Response Sent:
+  Emitted from a service after sending a response to a client.
+:Response Received:
+  Emitted from a client after receiving a response from a service.
+
+Request event messages shall be published to the hidden topic ``/SERVICE_NAME/_request_event``, where ``SERVICE_NAME`` is the fully-qualified name of the service.
+Similarly, response event messages shall be published to the hidden topic ``/SERVICE_NAME/_response_event``.
+Note that this implies that services must have unique names.
+
+By publishing service event messages to predetermined topics, tools and libraries are able to subscribe to these topics to inspect the flow of data between services.
+
+Service Event Definitions
+-------------------------
+
+For each service definition, ``my/srv/Foo``, we define two new ROS message types using the ROS IDL specification [2]_:
+
+:my/msg/Foo_RequestEvent:
+  This type communicates a request sent or request received event.
+
+  .. code-block::
+
+     # Indicates this request event was emitted from a client
+     uint8 CLIENT_SENT      = 0
+
+     # Indicates this request event was emitted from a service
+     uint8 SERVICE_RECEIVED = 1
+
+     # Whether this is a CLIENT_SENT or SERVICE_RECEIVED event.
+     uint8 request_type
+
+     # Timestamp for when the event occurred (sent or received time)
+     builtin_interfaces/msg/Time stamp
+
+     # Unique identifier for the client that sent the request
+     # Note, this is only unique for the current session
+     unique_identifier_msgs/msg/UUID client_id
+
+     # Sequence number for the request
+     # Combined with the client ID, this creates a unique ID for the request
+     int64 sequence_number
+
+     # The actual request content sent or received
+     # Setting this field is optional
+     my/srv/Foo_Request request
```

Comment:
I think this is worth considering. It would also allow us to use a single topic for all service events (e.g. we could include the service name in the message as well). A single topic makes it more predictable what topics to expect. With multiple topics, tools that want to interact with all service events first have to get a list of all services and then infer topic names to subscribe to.

I don't think it's unreasonable for us to provide libraries in C++ and Python to make working with service events more ergonomic. It will be slightly more work to support in other languages, which would have to provide their own helpers  for deserializing.

Reply:
> It would also allow us to use a single topic for all service events (e.g. we could include the service name in the message as well)

I prefer one events topic per service though, which seems to be the more common case to introspect.

But if the default is one events topic per service and the message definition provides all the needed information (including the service name and type), users could remap all (or some of) the service events topics to the same one if they wanted (and we could maybe have an option to do that more easily).

Reply:
Having separate topics per service makes it easier for tools to selectively introspect (or record) a service. If all are on the same topic they'd have to implement a filter on the content of the event message.